Audit bug fixes 🚫🐛

kautukkundan opened the pull request:
**What does this pull request do? Explain your changes. (required)**
<!-- A clear and concise description of what this pull request does. -->
Fixes bugs found in [audit](https://github.com/code-423n4/2022-01-livepeer-findings/issues) excluding gas optimization changes.

Note for reviewer -
Each commit links to an individual issue so it is recommended to go over each commit hash separately and then review it

**Specific updates (required)**
<!--- List out all significant updates your code introduces -->
check commit history

**How did you test each of these updates (required)**
<!-- A detailed description of how you tested your code changes. Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->
ran test suite

**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] README and other documentation updated
- [x] All tests using `yarn test` pass

yondonfu commented on test/unit/L1/l1LPTGateway.test.ts:
```diff
@@ -299,12 +317,89 @@ describe('L1 LPT Gateway', function() {
+          await expect(tx).to.be.revertedWith('WRONG_ETH_VALUE');
+        });
+
+        it('should revert when supplied ETH is more than submission cost', async function() {
```
```suggestion
        it('should revert when supplied ETH is more than expected ETH', async function() {
```

kautukkundan replied:
addressed in df1be67

yondonfu commented on test/unit/L1/l1LPTGateway.test.ts:
```diff
@@ -299,12 +317,89 @@ describe('L1 LPT Gateway', function() {
+          await expect(tx).to.not.be.reverted;
+        });
+
+        it('should revert when supplied ETH is less than submission cost', async function() {
```
```suggestion
        it('should revert when supplied ETH is less than expected ETH', async function() {
```

kautukkundan replied:
addressed in df1be67

yondonfu commented on test/unit/L2/delegatorPool.test.ts:
```diff
@@ -176,10 +184,39 @@ describe('DelegatorPool', function() {
 
+        it('fails if requested stake exceeds initial stake', async () => {
+          const unclaimedTokens = 10;
+          await delegatorPool
```
I suggest adding the following expect line before these test cases:

```
expect(await delegatorPool.initialStake()).to.be.equal(stake)
```

in order to convey to the reader that `stake` is equal to the `initialStake` tracked by the DelegatorPool so that in the second test case below it is clear why claiming `unclaimedTokens + 1` would trigger a revert (i.e. since the first claim was for `initialStake - unclaimedTokens`).

kautukkundan replied:
addressed in b2b88fe

yondonfu commented on test/unit/L1/l1Migrator.test.ts:
```diff
@@ -920,6 +914,30 @@ describe('L1Migrator', function() {
+        const amount = 200;
+        const seqNo = 7;
+
+        bridgeMinterMock.withdrawETHToL1Migrator.returns(amount);
```
I think we can drop these two lines since `migrateLPT()` will never call these functions since a modifier will trigger a revert first.

kautukkundan replied:
addressed in d865442

yondonfu reviewed (changes requested):
Generally looks good. Just a few suggestions.

When you address the comments I suggest using fixup commits so before merging you can easily autosquash the fixup commits into the previous commits that they should be modifying. For example, suppose we want to fixup 0286aac2887956015b2c2de2a8a5e9c73d9e4aa4 with a change to the tests. We could make the relevant changes, stage them and then do:

```
git commit --fixup 0286aac2887956015b2c2de2a8a5e9c73d9e4aa4
```

which will create a commit that is marked as a fixup commit. Then, when we rebase later on the fixup commit will be automatically combined with the target commit.

Refer to [this article](https://thoughtbot.com/blog/autosquashing-git-commits) for more information on fixup commits and autosquashing as tools to help maintain a clean commit history which can be especially helpful in this scenario since we want to make it clear which changes are addressing which audit findings.

kautukkundan commented:
Unsure why block-fixup check is failing

yondonfu reviewed (commented):
Changes look good to me!

> Unsure why block-fixup check is failing

That GH action will fail if there are any fixup commits that haven't been squashed yet.

So, as the last step before merging let's squash the fixup commits (you'll want to force push after completing the rebase to autosquash those commits).

kautukkundan commented:
Squashed, however I guess since I added the 'fixes' keyword in the commits, it will close the linked issue in the contest repo while judgement is still in progress. 

I think it will be better to use "squash and merge" while merging and change the _fixes_ keyword to _addresses_ or _related_. WDYT?

yondonfu commented:
> I think it will be better to use "squash and merge" while merging and change the fixes keyword to addresses or related. WDYT?

My suggestion is to:

1. Drop the "fixes" keyword altogether from all the commit messages
2. Fix the conflict
3. Merge using "rebase and merge". The problem with "squash and merge" here is that it will collapse all the commits into a single one. While that single commit will link to all the relevant issues it will be more difficult to isolate the specific changes that were made for specific issues. I think the ability to isolate the specific changes for specific issues into their own commits is a nice property.

Once 1 & 2 is complete should be ready for approval.

yondonfu reviewed (approved):
LGTM!
